### PR TITLE
Test that FSDP2 works with cuda graphs.

### LINKED ISSUE #164264: Cuda graph support for FSDP2 is lacking.
### 🚀 The feature, motivation and pitch

I was asked recently to investigate allowing use of FSDP2 with cuda graph in pytorch.

I discovered several issues:

1. torch.chunk_cat, which is used in FSDP2, allocates pinned host memory. That is currently unsafe in cuda graph capture, but I have a fix to this that I will upload shortly.
2. make_graphed_callables does not support running with models with any hooks. It says you can add hooks to a torch.nn.Module *after* having called make_graphed_callables on it, but this is clumsy because you need the hooks to all-gather all of your weights in order to warmup. Warmup is the only safe way to guarantee a  workload can be captured to a cuda graph today. If you could somehow add the hooks for the sake of warmup, remove the hooks for the sake of make_graphed_callables, and then re-add the hooks, I suppose you could warmup successfully. However, make_graphed_callables is an old API that came out when cuda graphs were first added to pytorch. I personally do not think that [the requirement that a module not have hooks](https://github.com/pytorch/pytorch/blob/84e1cd73929c9935d8381cd7e549199ecf09ff10/torch/cuda/graphs.py#L385-L397) is overly cautious. We don't need it. I am investigating making a new variant of make_graphed_callables that drops this requirement.
3. At least once, a stream fails to join against the capturing stream during stream capture: https://github.com/pytorch/pytorch/blob/1981ed4f60ba54e54373bb44381e301a4b8f594b/torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py#L247 This means this work is not recorded inside the cuda graph, causing a mismatch between graph replay and eager execution.

Unfortunately, several workloads using FSDP2 can become host latency bounded, so cuda graph support is something reasonable to want for it.

I can take care of these issues myself, but I wanted to document them so that people are aware. I also don't personally know the FSDP2 maintainers, so I wanted to make sure they were aware as well.

### Alternatives

_No response_

### Additional context

_No response_

cc @ezyang @gchanan @kadeng @msaroufim @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @dcci @mcarilli @eellison @penguinwu @BoyuanFeng @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @muchulee8 @amjames @chauhang @aakhundov @coconutruben @zou3519

### PR DESCRIPTION
**Pull request by galv**
I initially wrote in #164264 that there was a missing wait_stream() call to put a stream into stream capture mode, but surprisingly since I made that issue the problem has been fixed. I was not able to locate the exact commit that coincidentally made that fix after a brief search. Since CachingHostAllocator supports memory allocation during stream capture since #167507, the purpose of this PR is simply to make sure that the support does not regress.

An important detail is that we need to make sure that cuda graph still overlaps the all-gather and reduce-scatter streams with computation streams. To check for that, I applied this patch:

```
diff --git a/test/distributed/_composable/fsdp/test_fully_shard_training.py b/test/distributed/_composable/fsdp/test_fully_shard_training.py
index c0831d87d7c..c0fecdf787d 100644
--- a/test/distributed/_composable/fsdp/test_fully_shard_training.py
+++ b/test/distributed/_composable/fsdp/test_fully_shard_training.py
@@ -1681,8 +1681,8 @@ class TestFullyShardCudaGraph(FSDPTest):
         device = torch.device(device_type.type, self.rank)
         torch.manual_seed(42)
         model = nn.Sequential(
-            nn.Linear(8, 8, bias=False),
-            nn.Linear(8, 8, bias=False),
+            nn.Linear(4096, 4096, bias=False),
+            nn.Linear(4096, 4096, bias=False),
         ).to(device)
         for param in model.parameters():
             dist.broadcast(param, src=0)
@@ -1694,7 +1694,7 @@ class TestFullyShardCudaGraph(FSDPTest):

         # warmup
         with torch.cuda.stream(stream):
-            input_tensor = torch.randn(4, 8, device=device)
+            input_tensor = torch.randn(4, 4096, device=device)
             output = model(input_tensor)
             output.sum().backward()
             model.zero_grad(set_to_none=True)
@@ -1711,7 +1711,7 @@ class TestFullyShardCudaGraph(FSDPTest):
             ]

         # equivalence check
-        with torch.cuda.stream(stream):
+        with torch.cuda.stream(stream), torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CPU, torch.profiler.ProfilerActivity.CUDA], record_shapes=True, profile_memory=True) as prof:
             for _ in range(2):
                 replay_input = torch.randn(4, 8, device=device)
                 ref_output = model(replay_input)
@@ -1726,6 +1726,8 @@ class TestFullyShardCudaGraph(FSDPTest):
                 for graph_grad, ref_grad in zip(static_output_grads, ref_grads):
                     self.assertTrue(torch.equal(graph_grad, ref_grad))
                 model.zero_grad(set_to_none=True)
+                prof.step()
+        prof.export_chrome_trace(f"two_layer_fully_shard_cudagraph_{self.rank}.json")

 if __name__ == "__main__":
```

I then inspection the json file manually to check for overlap.

Closes issue #164264

Fixes #164264

cc @mcarilli @ezyang @eellison @penguinwu @BoyuanFeng

### CODE REVIEW

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 1]
```diff
@@ -1667,5 +1668,65 @@ def _shard_placement_fn(param: nn.Parameter) -> Optional[Shard]:
             self.assertEqual(losses[0], losses[1])
 
 
+class TestFullyShardCudaGraph(FSDPTest):
+    @property
+    def world_size(self) -> int:
+        return 2
+
+    @skip_if_lt_x_gpu(2)
+    @unittest.skipIf(TEST_CUDA_GRAPH, "CUDA >= 11.0 or ROCM >= 5.3 required for graphs")
```
Comment (Skylion007) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py]:
```suggestion
    @unittest.skipIf(not TEST_CUDA_GRAPH, "CUDA >= 11.0 or ROCM >= 5.3 required for graphs")

**Review by ezyang — APPROVED** [review]

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 2]
```diff
@@ -1667,5 +1668,67 @@ def _shard_placement_fn(param: nn.Parameter) -> Optional[Shard]:
...
+            for _ in range(2):
+                replay_input = torch.randn(4, 8, device=device)
+                ref_output = model(replay_input)
+                ref_output.sum().backward()
+                ref_grads = [
+                    param.grad.detach().clone() for param in model.parameters()
+                ]
+
+                static_input.copy_(replay_input)
+                graph.replay()
```
Comment (ngimel) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py:1726]: lol if you attempted to do this for real you would be debugging for a long time why gradients are not accumulated, but for tests this is good enough

**Review by ngimel — APPROVED** [review]

**Review by BoyuanFeng — APPROVED** [review]

**Review by eellison — APPROVED** [review]

File: test/distributed/_composable/fsdp/test_fully_shard_training.py (hunk 2 above)
Comment (galv) [reply test/distributed/_composable/fsdp/test_fully_shard_training.py:1726]: For what it's worth, the idiom of doing `model.zero_grad(set_to_none=True)` before stream capture comes from the [original pytorch blog post on cuda graph](https://pytorch.org/blog/accelerating-pytorch-with-cuda-graphs/), so it wouldn't surprise me if most code out in the wild does this.

Very unfortunate for anyone who might be using cuda graphs this way and trying to do gradient accumulation over multiple minibatches 😬
